some changes: add some conf parameter,  add toleration for master taints, compatible with 0.14 parameter, version up docker base image

cw-sakamoto opened the pull request:
Hi, thanks for great product. 

I modified confd settings to make it possible to pass some parameters with environment variables.


Sorry, I incremented version +2 in Makefile, because I fixed two parts devided into two commit and create each version docker image(0.9.10, 0.9.11) before this PR.
(https://hub.docker.com/r/cwsakamoto/kube-fluentd/tags/)

Would you mind confirming parts I modified?

---

### FYI, abount retry_limit
In fluentd, ```retry_limit_disable``` is true or false value and if this parameter is true ```retry_limit``` is ignored.

- ```retry_limit_disable``` in fluentd code
  - https://github.com/fluent/fluentd/blob/v0.14.14/lib/fluent/compat/output.rb#L234
- ```retry_lmit_disable``` is converted to ```retry_forever``` and if ```retry_forever``` is true fluentd does not calculate retry_count
    - I showed the code for retry action in fluentd(0.14.14)
        - https://github.com/fluent/fluentd/blob/v0.14.14/lib/fluent/plugin/output.rb#L1111
        - https://github.com/fluent/fluentd/blob/v0.14.14/lib/fluent/plugin_helper/retry_state.rb#L20
        - https://github.com/fluent/fluentd/blob/v0.14.14/lib/fluent/plugin_helper/retry_state.rb#L87

cw-sakamoto committed:
add retry_lmit

cw-sakamoto committed:
add fluentd parameter for confd(flush_interval, disable_retry_limit)

cw-sakamoto committed:
modified Docker.template and Makefile

cw-sakamoto committed:
change docker repogitory

mumoshu commented on rootfs/etc/confd/templates/fluent.conf.tmpl:
```diff
@@ -224,8 +225,9 @@
   max_retry_wait 30
-  disable_retry_limit
+  disable_retry_limit {{getv "/fluentd/out/kubesys/retry/limit/disable"}}
+  retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
```
Although my idea would not be that of a well-trained fluentd user, I'd prefer something like:

```
{{if eq (getv "/fluentd/out/kubesys/retry/limit") "-1" -}}
disable_retry_limit true
{{else -}}
retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
{{end -}}
```

so that, imho, the configuration becomes a bit more idiomatic. I've seen a retry limit being "-1" meaning "retry forever" in other job queues and libraries.

cw-sakamoto replied:
@mumoshu Thanks for your comment!

>the configuration becomes a bit more idiomatic

I agreed, it's good!!

I do not understand how to write confd template,,,  I'd like to setting conf like this:

```
{{if le (getv "/fluentd/out/kubesys/retry/limit") 0 -}}
disable_retry_limit true
{{else -}}
retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
{{end -}}
```

Is it possible?

cw-sakamoto replied:
While watching [this page](https://github.com/kelseyhightower/confd/blob/master/docs/templates.md), I came up with something  like:

```
{{if exists "/fluentd/out/kubesys/retry/limit" }}
retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
{{else}}
disable_retry_limit true
{{end}}
```
How do you think ? 😄

mumoshu replied:
@cw-sakamoto How do you "unset" the retry/limit var to a value other than the default 50 then? If that's possible, it looks good to me :+1:

mumoshu replied:
@cw-sakamoto How do you "unset" the retry/limit var to a value other than the default 50 then? If that's possible, it looks good to me :+1:

cw-sakamoto replied:
@mumoshu 
Humm...I don't think "unset" this var 😮  Sorry, it was just idea...
Thanks for your considering 👍 

I thought about various cases, but after all I think that the original writing style is good.

so that:
- fluentd user doesn't get confused when setting parameters of kube-fluentd according to fluentd's original specification
  - fluentd user (maybe) knows ```disable_retry_limit``` is boolean value, and disable ```retry_limit``` parameter.
  - When user want to set ```diable_retry_limit```, the same as fluentd is easier to set this parameter.

mumoshu replied:
Concretely speaking, perhaps what you couldn't agree with was the "set retry/limit to -1 for retrying forever" part of my suggestion, right?

Then, I'd suggest:

```
{{if exists "/fluentd/out/kubesys/retry/forever" }}
retry_forever
{{else}}
retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
{{end}}
```

cw-sakamoto replied:
>Concretely speaking, perhaps what you couldn't agree with was the "set retry/limit to -1 for retrying forever" part of my suggestion, right?

Yes, I think your suggest becomes configuration more idiomatic and it is important.
But I'm concerned that user could get confused because kube-fluentd's parameter specification is different fluentd's original parameter one.

How do you think my suggestion? 
```
{{if (getv "/fluentd/out/kubesys/retry/forever")}}
retry_forever
{{else}}
retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
{{end}}
```

so that:
- ```"/fluentd/out/kubesys/retry/forever"``` is set to default ```false``` in Dockerfile
- if user want to use retry_forever, user set ```"/fluentd/out/kubesys/retry/forever"``` is other than ```false```
  - ```"/fluentd/out/kubesys/retry/limit"``` is determined if it is valid depending on ```"/fluentd/out/kubesys/retry/forever"```
    - this specification is almost the same to fluentd original one

But, retry_forever is 0.14 only parameter, and retry_limit is 0.12 parameter(retry_limit is converted to retry_max_times internally)... So I'll rewrite with 0.14 parameters.

Please wait ... 😃

mumoshu replied:
@cw-sakamoto Thanks for the suggestion 👍 

> if user want to use retry_forever, user set "/fluentd/out/kubesys/retry/forever" is other than false

Perhaps I'm missing something here but does it really work like that?

> ```
> {{if (getv "/fluentd/out/kubesys/retry/forever")}}
> retry_forever
> {{else}}
> retry_limit {{getv "/fluentd/out/kubesys/retry/limit"}}
> {{end}}
> ```

AFAIK, `getv` returns a string value and therefore, won't it always default to `{{if "false"}}` which would be considered `true` in practice? 

> So I'll rewrite with 0.14 parameters.

Great 👍

cw-sakamoto replied:
Thanks for your comment.

>AFAIK, getv returns a string value and therefore, won't it always default to {{if "false"}} which would be considered true in practice?

Sorry, it was collect as you said.  😿  I checked my local.
```"false"``` is not boolean, so condition is true. Trying it elsewhare, ```""``` is false and other(```"0"```, ```"false"```) is true.


I'll think about this problem while rewriting for 0.14.

mumoshu reviewed (changes requested):
@cw-sakamoto Thanks a lot for your efforts!
I especially like the detailed PR description ❤️ 

Regarding the code, I've left one thing about how `disable_retry_limit` could be configured.
Everything else LGTM 👍

cw-sakamoto committed:
add toleration for master taints

cw-sakamoto committed:
change fluentd parameters for 0.14

cw-sakamoto committed:
version up docker base image

cw-sakamoto commented:
@mumoshu 
Hi, apologize for lazy commit...
Additionally, I changed the following points:
- add toleration for master taints to daemonset file
- compatible with fluentd 0.14 parameters
- version up docker base image
  - for security
- version up fluentd
  - for improvement fluentd's logging output

In accordance with the above points, I changed the version to 0.9.12 and fluentd version 0.14.22, so Docker image tag is 0.14.22-0.9.12.

Would do you mind confirming my changes?

---
By the way, I fixed about parameters ```disable_retry_limit```(```retry_forever```)
```
{{if eq (getv "/fluentd/out/kubesys/retry/forever") "true" -}}
retry_forever true
{{else -}}
retry_max_times {{getv "/fluentd/out/kubesys/retry/limit"}
{{end -}}
```
How do you think about conditions?

mumoshu commented on fluentd.daemonset.yaml:
```diff
@@ -12,7 +12,13 @@ spec:
+        - operator: Exists
+          effect: NoExecute
+        - operator: Exists
+      dnsPolica: Default
```
nit: dnsPolicy

cw-sakamoto replied:
😢

cw-sakamoto replied:
I fixed in bcf8387

mumoshu commented on rootfs/etc/confd/conf.d/fluent.conf.toml:
```diff
@@ -2,10 +2,20 @@
+    "/fluentd/out/kubesys/buffer/flush/interval",
+    "/fluentd/out/kubesys/buffer/retry/max/interval",
+    "/fluentd/out/kubesys/retry/forever",
+    "/fluentd/out/kubesys/retry/limit",
```
Would you prefer to name it `retry/max/times`(and _RETRY_MAX_TIMES) instead of `retry/limit`(and _RETRY_LIMIT) following [the corresponding configuration key in fluentd v0.14](https://docs.fluentd.org/v0.14/articles/buffer-section#retries-parameters)?

cw-sakamoto replied:
Sorry for my mistake.

I'd like to correspond confd keys with fluentd(0.14) configuration.
I fixed this in 2b3f26a

cw-sakamoto committed:
modified confd keys (correspoding configuration in fluentd)

cw-sakamoto committed:
fixed typo

mumoshu commented:
@cw-sakamoto LGTM. Thank you very much for your efforts!
